File: CoinChange.java
import java.util.Arrays;

public class CoinChange {
    //322. Coin Change
    //https://leetcode.com/problems/coin-change/description/
    public int coinChange(int[] coins, int amount) {
        int[] res = new int[amount + 1];
        Arrays.fill(res, amount + 1); //Basically a way to indicate if the last value cannot be obtained with the coins used.
        res[0] = 0;

        for(int a = 0; a <= amount; a++){ //Going through every possible amount to save the sub-problem
            for(int c = 0; c < coins.length; c++){ //Iterate through the coins array
                if(a - coins[c] >= 0){ //If the difference between the amount and the coin used is not negative, add it into the dp array. 
                    res[a] = Math.min(res[a], 1 + res[a - coins[c]]); //Set the min between the number of coins in the current element and the number of coins from res[amount - value of the coin] + the previous coin obtained (1)
                }
            }
        }

        if(res[amount] == amount + 1){
            return -1;
        }else{
            return res[amount];
        }

    }
}

File: CoinChange.py
#322. Coin Change
#https://leetcode.com/problems/coin-change/
#Use tabluation. Itreate through evey possible amt to save the sub-problem via the coins arr
#If diff between the amt and coin used is not neg, add it into the sub-problem array.
#Set the min between the number of coins in the current element and the number of coins from res[amount - value of the coin] + the previous coin obtained (1)
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        res = [amount + 1] * (amount + 1)
        res[0] = 0

        for a in range(1, amount + 1):
            for coin in coins:
                if coin <= a:
                    sub_res = res[a - coin]
                    if sub_res != amount + 1 and sub_res + 1 < res[a]:
                        res[a] = sub_res + 1
        
        print(res)

        if res[amount] == amount + 1:
            return -1
        else:
            return res[amount]